restarter/hot-restarter.py

```diff
@@ -5,18 +5,69 @@
 import sys
 import time
 
+# The number of seconds to wait for children to gracefully exit after
+# propagating SIGTERM before force killing children.
+# NOTE: If using a shutdown mechanism such as runit's `force-stop` which sends
+# a KILL after a specified timeout period, it's important to ensure that this
+# constant is smaller than the KILL timeout
+TERM_WAIT_SECONDS = 30
+
 restart_epoch = 0
 pid_list = []
 
-def force_kill_all_children():
-  """ Iterate through all known child processes and force kill them. In the future we might consider
-      possibly giving the child processes time to exit but this is fine for now. If someone force kills
-      us and does not clean the process tree this will leave child processes around unless they choose
-      to end themselves if their parent process dies. """
+def term_all_children():
+  """ Iterate through all known child processes, send a TERM signal to each of
+  them, and then wait up to TERM_WAIT_SECONDS for them to exit gracefully,
+  exiting early if all children go away. If one or more children have not
+  exited after TERM_WAIT_SECONDS, they will be forcibly killed """
 
   # First uninstall the SIGCHLD handler so that we don't get called again.
   signal.signal(signal.SIGCHLD, signal.SIG_DFL)
 
+  global pid_list
+  for pid in pid_list:
+    print "sending TERM to PID={}".format(pid)
+    try:
+        os.kill(pid, signal.SIGTERM)
+    except:
+      print "error sending TERM to PID={} continuing".format(pid)
+
+  all_exited = False
+
+  # wait for 30 seconds for children to exit cleanly
+  retries = 0
+  while not all_exited and retries < TERM_WAIT_SECONDS:
+    pid_list_copy = list(pid_list)
+    for pid in pid_list_copy:
```

Comment:
I would just write `for pid in list(pid_list)`. I have a mild concern about this being O(n^2), but it's fine given how few children that exist in reality. It would probably be cleaner to model the `pid_list` as a `set`.

Reply:
I thought the reason for the copy is because we're removing elements from the list as we iterate. that's what's done in `sigchld_handler()`